packages/react-dom/src/__tests__/ReactMount-test.js

```diff
@@ -173,12 +482,632 @@ describe('ReactMount', () => {
         div,
       ),
     ).toWarnDev(
-      'Server: "This markup contains an nbsp entity:   server text" ' +
-        'Client: "This markup contains an nbsp entity:   client text"',
+      'Warning: Text content did not match. ' +
+        'Server: "This markup contains an nbsp entity:   server text" ' +
+        'Client: "This markup contains an nbsp entity:   client text"\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should warn when a hydrated element has extra props with non-null values', () => {
+    // See fixtures/ssr: ssr-warnForPropDifference
+
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <div>
+        <em>SSRMismatchTest default text</em>
+      </div>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div data-ssr-extra-prop={true} data-ssr-extra-prop-2={true}>
+          <em>SSRMismatchTest default text</em>
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Prop `data-ssr-extra-prop` did not match. ' +
+        'Server: null ' +
+        'Client: true\n' +
+        '    in div (at **)',
+    );
+  });
+
+  it('should not warn when a hydrated element has extra props explicitly set to null', () => {
+    // See fixtures/ssr: ssr-warnForPropDifference-null-no-warning
+
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <div>
+        <em>SSRMismatchTest default text</em>
+      </div>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div data-ssr-extra-prop={null} data-ssr-extra-prop-2={null}>
+          <em>SSRMismatchTest default text</em>
+        </div>,
+        div,
+      ),
+    ).toWarnDev([]);
+  });
+
+  it('should warn when a server element has extra props', () => {
+    // See fixtures/ssr: ssr-warnForExtraAttributes
+
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <div data-ssr-extra-prop={true} data-ssr-extra-prop-2={true}>
+        <em>SSRMismatchTest default text</em>
+      </div>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <div>
+          <em>SSRMismatchTest default text</em>
+        </div>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Extra attributes from the server: data-ssr-extra-prop,data-ssr-extra-prop-2\n' +
```

Comment:
TODO: Print the attribute names array nicer, with a space after comma. Now the default array toString is used implicitly.

Reply:
> TODO: Print the attribute names array nicer, with a space after comma. Now the default array toString is used implicitly.

Fixed in https://github.com/facebook/react/pull/13602/commits/9c8093f3516b695239620d128420e9870ec96787